### DISCUSSION TITLE
How the FelixConfiguration is being recreated and how to change its default values?

### DISCUSSION THREAD

**Discussion by PawelJablonski-Dynatrace** [projectcalico/calico#9147] [Q&A] [answered]
Hello Team,

I would like to understand the behavior of recreating FelixConfiguration resource. I have reinstalled calico/tigera to version v3.28.1 on my k8s cluster. By default it created following FelixConfiguration resource: 

```
% kubectl -n calico-system describe FelixConfiguration default
Name:         default
Namespace:    
Labels:       <none>
Annotations:  operator.tigera.io/bpfEnabled: false
              projectcalico.org/metadata:
                {"uid":"32bb45a5-ec0f-4916-babd-6f21fc89972b","generation":1,"creationTimestamp":"2024-08-09T10:09:35Z","managedFields":[{"manager":"opera...
API Version:  crd.projectcalico.org/v1
Kind:         FelixConfiguration
Metadata:
  Creation Timestamp:  2024-08-09T10:09:35Z
  Generation:          4
  Resource Version:    747891792
  UID:                 32bb45a5-ec0f-4916-babd-6f21fc89972b
Spec:
  Bpf Enabled:          false
  Bpf Log Level:        
  Floating I Ps:        Disabled
  Health Port:          9099
  Log Severity Screen:  Info
  Reporting Interval:   0s
  Route Table Range:
    Max:      99
    Min:      65
  Vxlan VNI:  4096
Events:       <none>
```

This configuration results in following log entry: 

```
[UnknownClass] felix/config_params.go 1043 Proceeding with `RouteTableRange` config option This field has been deprecated in favor of `RouteTableRanges` 
```

I have modified the spec entry locally and reapplied it via calicoctl to following values: 

```
spec:
  bpfEnabled: false
  bpfLogLevel: ""
  floatingIPs: Disabled
  healthPort: 9099
  logSeverityScreen: Info
  reportingInterval: 0s
  routeTableRanges:
  - max: 99
    min: 65
  vxlanVNI: 4096
```

This was accepted and after restart of calico-nodes the log was changed to: 

```
felix/config_params.go 1083: Both `RouteTableRanges` and deprecated `RouteTableRange` options are set. `RouteTableRanges` value will be given precedence.
```

When I have tried to describe the FelixConfiguration once again, I found that both options were present, 'routeTableRanges' and my previous 'routeTableRange' which I tried to remove. 

Then I tried to remove the FelixConfiguration manually and it was immediately recreated with its default values which I posted at the top of this post. 

What is recommendation for configuring this part to be upgrade-safe when the previous field will be removed? How to  get rid of the warning in the log?

**Comment by tmjd** [comment, marked as answer]
FelixConfiguration will have a default resource created by calico-node and/or operator.
If you are using AmazonVPC CNI plugins then the operator will set a default RouteTableRange, so I would guess it is the operator that keeps setting that field. (the operator also sets it if using GKE but due to the values you're seeing I'm assuming you're using AmazonVPC)
You should be able to see the CNI type in the Installation resource (`kubectl get installation default -o yaml`).

I've created https://github.com/tigera/operator/issues/3757 for this issue.
